proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2521,6 +2439,445 @@ lemma possibleSwitchTo_all_invs_but_ct_not_inQ':
+  apply (prop_tac "(t, EPSend) \<in> state_refs_of s epptr")
+   apply (clarsimp simp: state_refs_of_def)
+  apply (clarsimp simp: sym_refs_def dest!: invs_sym_refs)
+  apply (fastforce simp: state_refs_of_def)
+  done
+
+lemma in_recv_ep_queue_TCBBlockedRecv:
+  "\<lbrakk>kheap s epptr = Some (Endpoint (Structures_A.endpoint.RecvEP queue)); t \<in> set queue; invs s\<rbrakk>
+   \<Longrightarrow> (epptr, TCBBlockedRecv) \<in> state_refs_of s t"
+  apply (prop_tac "valid_ep (Structures_A.endpoint.RecvEP queue) s")
+   apply (fastforce simp: valid_objs_def valid_obj_def dest!: invs_valid_objs)
+  apply (clarsimp simp: state_refs_of_def valid_ep_def split: option.splits)
+  apply (intro conjI impI allI; (fastforce simp: obj_at_def)?)
+  apply (prop_tac "(t, EPRecv) \<in> state_refs_of s epptr")
+   apply (clarsimp simp: state_refs_of_def)
+  apply (clarsimp simp: sym_refs_def dest!: invs_sym_refs)
+  apply (fastforce simp: state_refs_of_def)
+  done
+
+lemma in_send_ep_queue_thread_state:
+  "\<lbrakk>ko_at' (Structures_H.endpoint.SendEP queue) epPtr s; t \<in> set queue;
+    sym_refs (state_refs_of' s); valid_objs' s\<rbrakk>
+   \<Longrightarrow> st_tcb_at' isBlockedOnSend t s"
+  apply (prop_tac "valid_ep' (SendEP queue) s")
+   apply (fastforce simp: valid_objs'_def valid_obj'_def obj_at'_def projectKOs)
+  apply (clarsimp simp: valid_ep'_def)
+  apply (prop_tac "(t, EPSend) \<in> state_refs_of' s epPtr")
+   apply (clarsimp simp: state_refs_of'_def obj_at'_def projectKOs)
+  apply (clarsimp simp: sym_refs_def obj_at'_def)
+  apply (fastforce simp: state_refs_of'_def tcb_bound_refs'_def st_tcb_at'_def get_refs_def2
+                         isBlockedOnSend_def obj_at'_def projectKOs
+                  split: if_splits thread_state.splits)
+  done
+
+lemma in_recv_ep_queue_thread_state:
+  "\<lbrakk>ko_at' (Structures_H.endpoint.RecvEP queue) epPtr s; t \<in> set queue;
+    sym_refs (state_refs_of' s); valid_objs' s\<rbrakk>
+   \<Longrightarrow> st_tcb_at' isBlockedOnReceive t s"
+  apply (prop_tac "valid_ep' (RecvEP queue) s")
+   apply (fastforce simp: valid_objs'_def valid_obj'_def obj_at'_def projectKOs)
+  apply (clarsimp simp: valid_ep'_def)
+  apply (prop_tac "(t, EPRecv) \<in> state_refs_of' s epPtr")
+   apply (clarsimp simp: state_refs_of'_def obj_at'_def projectKOs)
+  apply (clarsimp simp: sym_refs_def obj_at'_def)
+  apply (fastforce simp: state_refs_of'_def tcb_bound_refs'_def st_tcb_at'_def get_refs_def2
+                         isBlockedOnReceive_def obj_at'_def projectKOs
+                  split: if_splits thread_state.splits)
+  done
+
+lemma IdleEP_versus_SendEP_and_RecvEP:
+  "\<lbrakk>ep = Structures_A.endpoint.IdleEP \<Longrightarrow> P;
+    \<And>q. ep = Structures_A.endpoint.SendEP q \<or> ep = Structures_A.endpoint.RecvEP q \<Longrightarrow> P\<rbrakk>
+   \<Longrightarrow> P"
+  using Structures_A.endpoint.exhaust by auto
+
+lemma not_IdleEP_case_split:
+  "ep \<noteq> Structures_A.endpoint.IdleEP
```

Comment:
I think `Structures_A.IdleEP` would be clear enough. Isabelle will expand it out to its longest form when the name is not unique, but we don't have to follow that in input. Short is usually easier to read.

Reply:
Oh I didn't realise this. Cool!